driver/wnbd_dispatch.c

```diff
@@ -164,65 +147,53 @@ NTSTATUS WnbdDispatchRequest(
         default:
             Element->Srb->DataTransferLength = 0;
             Element->Srb->SrbStatus = SRB_STATUS_INVALID_REQUEST;
-            StorPortNotification(RequestComplete,
-                                 Element->DeviceExtension,
-                                 Element->Srb);
-            ExFreePool(Element);
-            InterlockedDecrement64(&DeviceInfo->Stats.UnsubmittedIORequests);
+            CompleteRequest(Device, Element, TRUE);
+            InterlockedDecrement64(&Device->Stats.UnsubmittedIORequests);
             continue;
         }
 
-        PWNBD_PROPERTIES DevProps = &DeviceInfo->UserEntry->Properties;
+        PWNBD_PROPERTIES DevProps = &Device->Properties;
 
         switch(RequestType) {
         case WnbdReqTypeRead:
             Request->Cmd.Read.BlockAddress =
                 Element->StartingLbn / DevProps->BlockSize;;
             Request->Cmd.Read.BlockCount =
-                Element->ReadLength / DevProps->BlockSize;
+                Element->DataLength / DevProps->BlockSize;
             break;
         case WnbdReqTypeWrite:
             Request->Cmd.Write.BlockAddress =
                 Element->StartingLbn / DevProps->BlockSize;
             Request->Cmd.Write.BlockCount =
-                Element->ReadLength / DevProps->BlockSize;
-            if (Element->ReadLength > Command->DataBufferSize) {
+                Element->DataLength / DevProps->BlockSize;
+            if (Element->DataLength > Command->DataBufferSize) {
                 // The user buffer must be at least as large as
                 // the specified maximum transfer length.
                 Element->Srb->SrbStatus = SRB_STATUS_INTERNAL_ERROR;
-                StorPortNotification(
-                    RequestComplete,
-                    Element->DeviceExtension,
-                    Element->Srb);
-                ExFreePool(Element);
-                InterlockedDecrement64(&DeviceInfo->Stats.UnsubmittedIORequests);
+                CompleteRequest(Device, Element, TRUE);
+                InterlockedDecrement64(&Device->Stats.UnsubmittedIORequests);
                 Status = STATUS_BUFFER_TOO_SMALL;
                 goto Exit;
             }
 
             PVOID SrbBuffer;
             if (StorPortGetSystemAddress(Element->DeviceExtension,
                                          Element->Srb, &SrbBuffer)) {
-                // TODO: consider moving this part to a helper function.
                 Element->Srb->SrbStatus = SRB_STATUS_INTERNAL_ERROR;
```

Comment:
Yep. The only issue that I see is the log level, we might want to avoid WNBD_LOG_ERROR in the hot path not only for performance reasons, but to avoid spamming the logs (especially if we're to use ETW).